db/range_del_aggregator.cc

```diff
@@ -224,6 +250,65 @@ class CollapsedRangeDelMap : public RangeDelMap {
     return parsed.sequence < iter_->second;
   }
 
+  bool ShouldDeleteRange(const Slice& start, const Slice& end,
+                         SequenceNumber seqno) {
+    ParsedInternalKey parsed_start;
+    if (!ParseInternalKey(start, &parsed_start)) {
+      assert(false);
+    }
+    ParsedInternalKey parsed_end;
+    if (!ParseInternalKey(end, &parsed_end)) {
+      assert(false);
+    }
+    if (ucmp_->Compare(parsed_start.user_key, parsed_end.user_key) > 0) {
+      return false;
+    }
+
+    auto iter = rep_.upper_bound(parsed_start.user_key);
+    if (iter == rep_.begin()) {
+      // before start of deletion intervals
+      return false;
+    }
+    --iter;
+    if (ucmp_->Compare(parsed_start.user_key, iter->first) < 0) {
+      assert(false);
+      return false;
+    }
+    // Loop looking for a tombstone that is older than the range sequence
+    // number, or we determine that our range is completely covered by newer
+    // tombstones.
+    for (; iter != rep_.end(); ++iter) {
+      if (ucmp_->Compare(parsed_end.user_key, iter->first) < 0) {
+        return true;
+      }
+      if (seqno >= iter->second) {
+        // Tombstone is older than range sequence number.
+        return false;
+      }
+    }
+    return false;
+  }
+
+  std::pair<RangePtr,SequenceNumber> GetTombstone(
+      const Slice& user_key, SequenceNumber seqno) {
+    auto iter = rep_.upper_bound(user_key);
+    if (iter == rep_.begin()) {
+      // before start of deletion intervals
+      return std::make_pair(RangePtr(nullptr, &iter->first), 0);
```

Comment:
I'm running a slightly different version of the code (tried to rebase locally). There was a problem here that I suspect exists in the original code as well: `iter->first` may be erased from the map when `AddTombstones` is called. Then the caller of `GetTombstone` can attempt to dereference a dangling pointer.

Reply:
oops, turns out this doesn't problem doesn't happen when running your original code. Still trying to understand why.

Reply:
I need to update this PR. We've replaced the use of a `std::pair` and `RangePtr` here with a `PartialRangeTombstone` struct. See https://github.com/cockroachdb/rocksdb/pull/12. It will be a few days before I can get to this update, though.

Reply:
Great, thanks for the pointer.